aten/src/ATen/core/type.cpp

```diff
@@ -351,7 +351,17 @@ MatchTypeReturn matchTypeVariables(
       type_env[vt->name()] = actual;
       return MatchTypeReturn::Success();
     } else if (auto unified = unifyTypes(it->second, actual)) {
-      type_env[vt->name()] = *unified;
+      // TODO: remove once tests that rely on old behavior have been fixed
```

Comment:
The loophole can also be patched by adding a specific overload for the append case. I think this is safer because it restricts the buggy back-compat behavior to just the append case seen in the wild and not to all optional str uses.

`append(List[str] s, str? elem)`